### ISSUE TITLE
Rule proposal: `prefer-class-fields`

### ISSUE THREAD

**Issue by MrHen** [sindresorhus/eslint-plugin-unicorn#314]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2430%20Funded-%2300A156.svg" />](https://oss.issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/314)

There is a superior pattern for class field declarations that can declutter constructors (often making them totally unnecessary.) The feature is still considered experimental but it works out of the box with babel.

Fail:

```js
class Foo {
  constructor() {
    this.foo = 'foo';
  }
}
```

Pass:

```js
class Foo {
  foo = 'foo';
}
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $30.00)

- [<img src='https://avatars.githubusercontent.com/u/1402241?v=4' alt='fregante' width=24 height=24> fregante](https://oss.issuehunt.io/u/fregante) ($30.00)

### Submitted pull Requests
- [#2512 Add `prefer-class-fields` rule](https://oss.issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/2512)
---

### Tips

- Checkout the [Issuehunt explorer](https://oss.issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://oss.issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by sindresorhus** [comment]
Sounds good, but I think we should wait until ESLint supports class fields.

---

Should it also handle:

```js
class Foo {
  constructor() {
    this._foo = 'foo';
  }
}
```

to

```js
class Foo {
  #foo = 'foo';
}
```

?

**Comment by MrHen** [comment]
I think a rule for private fields would be awesome, yeah. I don't think we can realistically fix those, however, because people abuse `_foo` access too regularly and it would very quickly break code.

**Comment by sindresorhus** [comment]
So you think private fields should be a separate rule?

---

> I don't think we can realistically auto-fix those, however, because people abuse _foo access too regularly and it would very quickly break code.

👍

**Comment by MrHen** [comment]
Yeah. As a general principle, I prefer rules to be have the smallest reasonable scope.

In this case, I can also see why a team might want to convert to class fields because of the fixer but not want to enable the private fields rule right away because they'd have to move a bunch of code around.

**Comment by sindresorhus** [comment]
Issue for private class fields: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/979

**Comment by sindresorhus** [comment]
Does ESLint support class fields yet?

**Comment by fregante** [comment]
It does now:

https://eslint.org/demo#eyJ0ZXh0IjoiY2xhc3MgQSB7XG5cdG5hbWUgPSAxXG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=

This rule would be pretty useful on Error constructors like:

```ts
class MyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MyError";
  }
}
```

👇


```ts
class MyError extends Error {
  name = "MyError"
}
```

**Comment by sindresorhus** [comment]
This is accepted.

**Comment by jimmywarting** [comment]
bump

**Comment by FRSgit** [comment]
FYI I've implemented this rule[ here](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2512), feel free to review & comment!
